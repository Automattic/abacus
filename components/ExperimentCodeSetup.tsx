import { createStyles, Link, makeStyles, Paper, Theme, Typography } from '@material-ui/core'
import React from 'react'

import Code from '@/components/Code'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(3),
    },
  }),
)

export default function ExperimentCodeSetup() {
  const classes = useStyles()
  return (
    <Paper className={classes.root}>
      <Typography variant='h4'>Experiment Code Setup</Typography>
      <Typography variant='subtitle1' gutterBottom>
        Connect this experiment to your code.
      </Typography>
      <br />
      <Typography variant='h5' gutterBottom>
        Using React
      </Typography>
      <Typography variant='body1'>
        <Link href='https://github.com/Automattic/wp-calypso/tree/master/client/components/experiment'>
          See this Readme
        </Link>
      </Typography>
      <br />
      <Typography variant='h5' gutterBottom>
        Using PHP
      </Typography>
      <Typography variant='body1'>
        See <Code>wp-content/lib/wpcom-abtest/wpcom-experiment.php</Code>
      </Typography>
    </Paper>
  )
}
