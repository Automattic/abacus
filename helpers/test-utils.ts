import mediaQuery from 'css-mediaquery'

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

export { createMatchMedia }
