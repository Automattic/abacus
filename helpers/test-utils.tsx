import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import mediaQuery from 'css-mediaquery'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}
