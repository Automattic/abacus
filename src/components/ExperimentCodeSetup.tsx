import { Box, createStyles, Link, makeStyles, Paper, Tab, Tabs, Theme, Typography } from '@material-ui/core'
import React, { useState } from 'react'
import Highlight from 'react-syntax-highlighter'

import { ExperimentFull } from '../lib/schemas'

interface TabPanelProps {
  children?: React.ReactNode
  index: number
  value: number
}

function TabPanel(props: TabPanelProps) {
  const { children, value, index, ...other } = props

  return (
    <div
      role='tabpanel'
      hidden={value !== index}
      id={`vertical-tabpanel-${index}`}
      aria-labelledby={`vertical-tab-${index}`}
      {...other}
    >
      {value === index && <Box p={3}>{children}</Box>}
    </div>
  )
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(3),
    },
  }),
)

export default function ExperimentCodeSetup({ experiment }: { experiment: ExperimentFull }): JSX.Element {
  const classes = useStyles()
  const [tabValue, setTabValue] = useState(0)

  const defaultVariation = experiment.variations.flatMap((x) => (x.isDefault ? x.name : null))[0] ?? 'control'
  const treatmentVariation = experiment.variations.flatMap((x) => (x.isDefault ? null : x.name))[0] ?? 'treatment'

  const react = `import React from 'react';
import Experiment, {
    DefaultVariation,
    Variation,
    LoadingVariations,
} from 'client/components/experiment';
function SomeExperiment( props ) {
    return (
        <Experiment name={ ${experiment.name} }>
            <DefaultVariation name="${defaultVariation}">
                Shows this variation when:
                <ol>
                    <li>The user doesn't have a variation assigned to them.</li>
                    <li>The experiment doesn't exist.</li>
                    <li>Before the experiment goes live</li>
                </ol>
            </DefaultVariation>
            <Variation name="${treatmentVariation}">Show variation B if the user is assigned this variation.</Variation>
            <LoadingVariations>
                Show this when we don't know the variation because we haven't called the API yet to get
                variations.
                <br />
                Protip: If this experiment is in the middle or end of a flow, try putting
                <code>&lt;QueryExperiments&gt;</code> at the beginning of your flow to ensure the variation
                assignments are preloaded.
            </LoadingVariations>
        </Experiment>
    );
}
`

  const wordpressPhp = `<?php

require_lib( 'wpcom-abtest/wpcom-experiment' );
$experiment = WPCOM_Experiment::instance( '${experiment.name}' );
$experiment->set_user( get_current_user_id(), $_COOKIE['tk_ai'] );
switch( $experiment->get_variation() ) {
    case '${treatmentVariation}':
        // do something with this variation
        break;
    default:
        // handle the default/control case
        break;
}
`

  const wordPressLanding = `<?php

require_once __DIR__ . '/../wp-content/lib/wpcom-abtest/wpcom-experiment-client.php';
$experiment = WPCOM_Experiment_Client::instance( '${experiment.name}' );
$experiment->set_user( null, $_COOKIE['tk_ai'] );
switch( $experiment->get_variation() ) {
    case '${treatmentVariation}'
        // do something with this variation
        break;
    default:
        // handle the default/control case
        break;
}
`
  return (
    <Paper className={classes.root}>
      <Typography variant='h4'>Experiment Code Setup</Typography>
      <Typography variant='subtitle1' gutterBottom>
        Connect this experiment to your code.
      </Typography>
      <br />
      <Tabs centered value={tabValue} onChange={(e, newValue) => setTabValue(newValue)}>
        <Tab label={'React'} />
        <Tab label={'WordPress.com'} />
        <Tab label={'Logged Out Homepages'} />
        <Tab label={'Raw HTTP'} />
      </Tabs>
      <TabPanel value={tabValue} index={0}>
        <Typography variant={'h5'}>Example Code</Typography>
        <Highlight language='js'>{react}</Highlight>
        <Typography variant={'h5'}>Documentation</Typography>
        <Link href='https://github.com/Automattic/wp-calypso/tree/master/client/components/experiment'>
          See this Readme
        </Link>
      </TabPanel>
      <TabPanel index={1} value={tabValue}>
        <Typography variant={'h5'}>Example Code</Typography>
        <Highlight language='php'>{wordpressPhp}</Highlight>
      </TabPanel>
      <TabPanel index={2} value={tabValue}>
        <Typography variant={'h5'}>Example Code</Typography>
        <Highlight language='php'>{wordPressLanding}</Highlight>
      </TabPanel>
      <TabPanel index={3} value={tabValue}>
        <Typography variant={'h5'}>Example curl</Typography>
        <Highlight language={''}>
          {`curl 'https://public-api.wordpress.com/wpcom/v2/experiments/0.1.0/assignments/${experiment.platform}?anon_id=123&user_id=1&_locale=en-US&geo=NL&_envelope=1'`}
        </Highlight>
        <Typography variant={'h5'}>Example Response</Typography>
        <Highlight langue={'HTTP'}>
          {`HTTP/2 200 
server: nginx
date: Wed, 14 Oct 2020 13:34:15 GMT
content-type: application/json; charset=UTF-8
vary: Accept-Encoding
x-hacker: Oh, Awesome: I/Opossum
x-robots-tag: noindex
link: <https://public-api.wordpress.com/>; rel="https://api.w.org/"
x-content-type-options: nosniff
access-control-expose-headers: X-WP-Total, X-WP-TotalPages, Link
set-cookie: ${experiment.name}=${treatmentVariation}; expires=Wed, 04-Nov-2020 00:00:00 GMT; Max-Age=1765545; path=/; domain=.wordpress.com; secure; SameSite=None
vary: Origin
access-control-allow-headers: Authorization, Content-Type
x-ac: 2.dfw _dfw 
strict-transport-security: max-age=15552000

{
    "body": {
        "variations": {
            "${experiment.name}": "${treatmentVariation}"
        },
        "ttl": 3600
    },
    "status": 200,
    "headers": {
        "Allow": "GET"
    }
}`}
        </Highlight>
      </TabPanel>
    </Paper>
  )
}
