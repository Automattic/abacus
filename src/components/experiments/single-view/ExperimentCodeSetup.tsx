import { createStyles, Link, makeStyles, Paper, Theme, Typography } from '@material-ui/core'
import React from 'react'

import Code from 'src/components/general/Code'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(3),
    },
  }),
)

export default function ExperimentCodeSetup(): JSX.Element {
  const classes = useStyles()
  return (
    <Paper className={classes.root}>
      <Typography variant='h4'>Experiment Code Setup</Typography>
      <br />
      <Typography variant='body1'>
        See{' '}
        <Link
          href="https://github.com/Automattic/experimentation-platform/wiki/Experimenter's-Guide#writing-the-controlvariant-code-experiences"
          rel='noopener noreferrer'
          target='_blank'
          underline='always'
        >
          the wiki
        </Link>{' '}
        for platform-specific instructions.
      </Typography>
      <br />
      <Typography variant='body1'>
        When testing manually, note that <strong>changes may take up to ten minutes to propagate</strong> due to{' '}
        <Link
          href="https://github.com/Automattic/experimentation-platform/wiki/Experimenter's-Guide#the-file-system-cache"
          rel='noopener noreferrer'
          target='_blank'
          underline='always'
        >
          the file system assignment cache
        </Link>
        . As specified in the wiki, you will need to run <Code>svn up</Code> to update your sandbox copy of the cache to
        reflect the latest changes.
      </Typography>
    </Paper>
  )
}
