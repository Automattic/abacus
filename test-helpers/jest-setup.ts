import { format } from 'util'

// Throw errors on any console logging.
// Adapted from https://github.com/facebook/jest/issues/6121#issuecomment-529591574
;['error', 'info', 'log', 'warn'].forEach((funcName) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(global.console as any)[funcName] = function (message?: any, ...optionalParams: any[]) {
    throw new Error(format(message, optionalParams))
  }
})
