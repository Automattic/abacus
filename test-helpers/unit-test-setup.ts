/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

// @ts-ignore
window.HTMLCanvasElement.prototype.getContext = jest.fn()
