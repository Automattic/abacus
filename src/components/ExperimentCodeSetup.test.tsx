import React from 'react'

import { render } from '@/test-helpers/test-utils'

import ExperimentCodeSetup from './ExperimentCodeSetup'

test('renders as expected', () => {
  const { container } = render(<ExperimentCodeSetup />)

  expect(container).toMatchSnapshot()
})
