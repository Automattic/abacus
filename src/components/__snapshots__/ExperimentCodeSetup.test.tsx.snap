// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4"
    >
      Experiment Code Setup
    </h4>
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
    >
      Connect this experiment to your code.
    </h6>
    <br />
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
    >
      Using React
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      <a
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/Automattic/wp-calypso/tree/master/client/components/experiment"
      >
        See this Readme
      </a>
    </p>
    <br />
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
    >
      Using PHP
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      See 
      <code
        class="MuiTypography-root makeStyles-root-2 MuiTypography-body1"
      >
        wp-content/lib/wpcom-abtest/wpcom-experiment.php
      </code>
    </p>
  </div>
</div>
`;
